File: week12/20/pr3_Battleships in a Board.cpp
using namespace std;
#include <vector>

class Solution {
public:
    int countBattleships(vector<vector<char>>& board) {
        int ans = 0;
        int m = board.size();
        int n = board[0].size();

        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(board[i][j]=='X'){
                    if((i==0 || board[i-1][j] =='.') && (j==0 || board[i][j-1]=='.')) ans++;
                    // 위 or 왼쪽이 없거나 '.'이면 전함 cnt증가
                }
            }
        }
        return ans;
    }
};

File: week12/20/pr3_Battleships in a Board.py
from typing import List
class Solution:
    def countBattleships(self, board: List[List[str]]) -> int:
        visited = []
        direction = [[0,1],[0,-1],[-1,0],[1,0],[0,0]]
        m = len(board)
        n = len(board[0])
        # find X and make all the neighbors of it visited
        def helper(i:int,j:int) -> None :
            for dx,dy in direction:
                new_i = i+dx
                new_j = j+dy
                if 0<=new_i<m and 0<=new_j<n and board[new_i][new_j] == "X" and (new_i,new_j) not in visited:
                    visited.append((new_i,new_j))
                    helper(new_i,new_j)
        cnt = 0
        for i in range(m):
            for j in range(n):
                if board[i][j] == "X" and (i,j) not in visited:
                    cnt+=1
                    helper(i,j)
        return cnt